---
title: "Scoring Patients"
author: "Sarah Killcoyne"
date: "22 June 2017"
output:
  html_document: 
    toc: yes
  word_document: default
---


```{r setup, include=FALSE}
library(ggplot2)
library(ggbio)
library(scales)
library(GGally)
library(plyr)
library(RColorBrewer)
library(dplyr)
library(pander)
library(ggdendro)
library(reshape2)
library(GenomicRanges)
library(gridExtra)
library(tibble)

source('lib/load_patient_metadata.R')

data = '~/Data/Ellie'

data.files = list.files(paste(data, 'QDNAseq',sep='/'), full.names=T)
plot.dir = paste(data, 'Analysis/multipcf_plots_fitted_perPatient', sep='/')

if (length(list.files(plot.dir)) <= 0)
  stop(paste("No analysis files found in", plot.dir ))

## Patient info file
patient.file = grep('/All_patient_info.xlsx', data.files, value=T)
if (length(patient.file) != 1)
  stop(paste("Missing/too many patient info file(s) in", data))
demo.file = grep('/Demographics_full.xlsx', data.files, value=T)

dataset = 'All'
info = read.patient.info(patient.file, demo.file, set=dataset)
removed.samples = info$removed
patient.info = info$info


#head(patient.info)

patient.info = arrange(patient.info, Status, Patient, Endoscopy.Year, Pathology)

sum.patient.data = summarise.patient.info(patient.info)
table(sum.patient.data$Status)

#patient.info = subset(patient.info, Patient %in% sum.patient.data$Patient & Batch.Name %in% levels(patient.info$Batch.Name)[1:5])

#pander(arrange(patient.data, Status), justify='left', caption='Patient information')
```


# Description of '`r dataset`' data set

```{r cors, echo=F, message=T, warning=F, fig.height=6, fig.width=6, fig.align='left', include=F}
## Cluster the samples by coverage then pull out regions

## Thresholds
gamma2 = 250
min.probes=67 # equiv to 1MB
sd.threshold = 0.08  # cov by numbers of probes
gain.threshold = 1.1; loss.threshold = 0.9

cachefile = paste(data, '/Analysis/', dataset, '_patients.Rdata', sep='')
if ( file.exists(cachefile) ) {
  load(cachefile, verbose=T)
} else {
  patient.data = list(); lowprobes = c()
  
  for (patient.name in sum.patient.data$Hospital.Research.ID  )  {
    message(patient.name)
    patient.plot.dir = paste(plot.dir, patient.name, sep='/')
  
    if (!file.exists(paste(patient.plot.dir, "/",patient.name,"_segmentedCoverage_fitted_gamma",gamma2,".txt",sep=""))) {
      warning(paste("Missing segmented file for", patient.name))
      next
    }
    
    segvals = read.table(paste(patient.plot.dir, "/",patient.name,"_segmentedCoverage_fitted_gamma",gamma2,".txt",sep=""),sep="\t",stringsAsFactors=F,header=T)
    head(segvals)
  
    missing = c( setdiff(colnames(segvals)[-(1:5)], subset(patient.info, Hospital.Research.ID == patient.name)$Samplename),
                 setdiff(subset(patient.info, Hospital.Research.ID == patient.name)$Samplename, colnames(segvals)[-(1:5)]) )
    
    mr = with(removed.samples, which(Hospital.Research.ID == patient.name & Samplename %in% missing))
    if (length(mr) > 0) {
      missing = missing[-which(missing %in% removed.samples[mr,]$Samplename)]
      segvals = segvals[,-which(colnames(segvals) %in% removed.samples[mr,]$Samplename) ]
    }

    if (length(missing) > 0) {
      warning(paste("Missing", length(missing), "sample(s) of", length(subset(patient.info, Patient == patient.name)$Samplename), "in patient", patient.name))
      #next
    }
  
    probes = length(which(segvals$n.probes < min.probes))
    message( paste(probes, ' probes (', round(probes/nrow(segvals),2)*100, '%)',' below the minimum probe count (',min.probes,')', sep=''))
    if (probes/nrow(segvals) >= 0.75) lowprobes = c(lowprobes, patient.name)
    
    segvals = segvals[segvals$n.probes >= min.probes,]
    dim(segvals)
  
    # ORDER segvals by endoscopy year etc
    cur.pt =  arrange( subset(patient.info, Hospital.Research.ID == patient.name), Endoscopy.Year, Pathology)
    segvals = segvals[, c(colnames(segvals[1:5]), intersect(colnames(segvals), cur.pt$Samplename))]
      
    if (nrow(segvals) < 100) {
      warning(paste(patient.name, 'has fewer than 100 genomic regions with segmented values:', nrow(segvals)))
      #next
    }
    
    patient.data[[patient.name]][['seg.vals']] = segvals
    patient.data[[patient.name]][['info']] = cur.pt
  
    sample.cols = grep('^D\\d+_D\\d+', colnames(segvals))
  
    if (length(sample.cols) > 0) {
      if (length(sample.cols) > 2) {  
        HC = hclust(dist(t(segvals[,sample.cols])))
        gg1 = ggdendrogram(HC, rotate=T) + labs(title = paste(patient.name,": fitted segment coverage"),x="")
        #print(gg1)
      }
  
      # Normalize  (value-mean(value))/sd(value)
      normalised.segvals = segvals[,sample.cols]
      if (length(sample.cols) > 1) {  
        for(c in 1:nrow(normalised.segvals)) 
          normalised.segvals[c,] = (normalised.segvals[c,]-mean(unlist(normalised.segvals[c,])))/sd(unlist(normalised.segvals[c,]))
        patient.data[[patient.name]][['norm.seg.vals']] = normalised.segvals
        
        HC = hclust(dist(t(normalised.segvals)))
        gg2 = ggdendrogram(HC, rotate=T) + labs(title=paste(patient.name,": normalised fitted segment coverage", x=""))
        #print(gg2)
  
        no.samples = ncol(normalised.segvals)
        no.rows = ceiling(no.samples/3)
  
        gg3 = ggplot( melt(normalised.segvals), aes(x=value) ) + 
          facet_wrap(~variable,scales = "free_x") + 
          geom_histogram( fill=I("blue"),col=I("lightblue"),binwidth=0.3) + labs(title=paste(patient.name,'normalised segment coverage'), x='', y='')
      }
        
      # Variable Regions can only be called relative to each sample so if there's only one sample we can't make a call
      if (length(sample.cols) > 1) {
        sds = apply(segvals[,sample.cols],1,sd)
        means = apply(segvals[,sample.cols],1,mean)
        CofV = sds/means
      
        variable.region.indices = which(sds >= sd.threshold & segvals$n.probes >= min.probes)
        variable.regions = segvals[variable.region.indices, intersect(patient.data[[patient.name]]$info$Samplename, colnames(segvals))] 
        patient.data[[patient.name]][['variable.regions']] = cbind(segvals[variable.region.indices,c(1:5)],variable.regions )
      }
      
      patient = arrange(subset(patient.info, Hospital.Research.ID == patient.name),  Endoscopy.Year, Pathology, Samplename)
    }
  }
save(patient.data, file=cachefile)
}
print(paste("Missing patients:", paste(setdiff(sum.patient.data$Hospital.Research.ID, names(patient.data)), collapse=', ')))

sum.patient.data = subset(sum.patient.data, Hospital.Research.ID %in% names(patient.data))
table(sum.patient.data$Status)


## Check that all of the samples are there
patient.info = arrange(patient.info, Status, Patient, Hospital.Research.ID, Endoscopy.Year,Samplename)
msgs = sapply(names(patient.data), function(pt) {
  missing = which(!patient.data[[pt]]$info$Samplename %in% colnames(patient.data[[pt]]$seg.vals))
  if (length(missing) > 0) {
    msg = paste(pt,": Missing samples - ", paste(patient.data[[pt]]$info$Samplename[missing], collapse=', ') , sep='')
    msg = paste(msg, "\nFrom batch: ", paste(unique(patient.data[[pt]]$info$Batch.Name[missing]), collapse=', '))
    warning(msg)
  }
})
for (n in msgs) {
  if (!is.null(n))
    print(n)
}

example.pt = which(with(sum.patient.data, Status == 'P' & total.samples > 2))[5]
key =  sum.patient.data[example.pt, 'Hospital.Research.ID' ]
example.pt.id = sum.patient.data[example.pt, 'Patient']
## Note the number of probes per patient/sample
```
\pagebreak

This data was generated in the following pipeline:

* Alignment using QDNAseq
* Filter excluded regions
* Joint segmentation per patient across all patient-derived-samples using PCF  (copynumber::multipcf method)

The result looks like two data matrices:

1. Segmented values (progressor): `r example.pt` 
```{r}
head(patient.data[[ key ]]$seg.vals)
```

```{r echo=F, warning=F, message=F, fig.height=10, fig.width=10}
ggplot( melt(patient.data[[key]]$seg.vals[,-(1:5)]), aes(x=value) ) + 
        facet_wrap(~variable,scales = "free_x") + 
        geom_histogram( fill=I("blue"),col=I("lightblue"), binwidth=0.03) + 
  labs(title=paste('Patient', example.pt.id,' segmented value distribution'), x='', y='')
```

2. Row (e.g. per-region across timepoints) normalized values from the matrix above
```{r}
head(patient.data[[key]]$norm.seg.vals)
```

```{r echo=F, warning=F, message=F, fig.height=10, fig.width=10}
ggplot( melt(patient.data[[key]]$norm.seg.vals), aes(x=value) ) + 
        facet_wrap(~variable,scales = "free_x") + 
        geom_histogram( fill=I("blue"),col=I("lightblue"), binwidth=0.3) + labs(title=paste('Patient', example.pt.id,'normalized segmented value distribution'), x='', y='')

```

Patients do not share the same chromosomal regions, though they can be overlapped, or the same timepoints. So all analyses have been conducted on a per-patient basis.

```{r general, warning=F, echo=F, message=F}
gen = do.call(rbind, lapply(patient.data, function(df) {
  cbind('n.segments'=nrow(df[['seg.vals']]), 
        'median.seg.len'=median(df$seg.vals[,c('end.pos')] - df$seg.vals[,c('start.pos')]),
        'mean.seg.len'=mean(df$seg.vals[,c('end.pos')] - df$seg.vals[,c('start.pos')]), 
        'sd.seg.len'=sd(df$seg.vals[,c('end.pos')] - df$seg.vals[,c('start.pos')]))
}))
rownames(gen)  = names(patient.data)

sum.patient.data = merge(sum.patient.data, gen, by.x='Hospital.Research.ID', by.y='row.names')
sum.patient.data = arrange(sum.patient.data, Status, total.samples)
sum.patient.data$years = with(sum.patient.data, paste(start.year, end.year, sep='-'))

ggplot(sum.patient.data) + geom_boxplot( aes(y=median.seg.len, x=Status, group=Status, fill=Status) ) +
  labs(title="Median segment length by patient status")
```

`r pander(sum.patient.data[,c('Patient','Status','years','total.samples','highest.path','n.segments', 'median.seg.len')], justify='left')`

There's no significant difference between the length of segments in progressors vs non.  There's a larger spread of lengths in progressors, but the means are nearly identical.

```{r echo=F, message=F, warning=F, fig.height=15, fig.width=10, eval=F}
# Segmented Value Distributions
lapply(patient.data, function(pt) {
  name = pt$info$Patient[1]
  status = pt$info$Status[1]
  
  g1 = ggplot( melt(pt[['seg.vals']][,-(1:5)]), aes(x=value) ) + 
      facet_wrap(~variable,scales = "free_x") + 
      geom_histogram( fill=I("blue"),col=I("lightblue"), binwidth=0.03) + 
      labs(title='segmented value distribution', x='', y='')

  
  g2 = ggplot( melt(pt[['norm.seg.vals']]), aes(x=value) ) + 
      facet_wrap(~variable,scales = "free_x") + 
      geom_histogram( fill=I("blue"),col=I("lightblue"), binwidth=0.3) + 
      labs(title='normalized segmented value distribution', x='', y='')

  do.call('grid.arrange', list(g1,g2, top=paste('Patient', name, status, sep=" - "), ncol=2))
})

```

# Variable Regions

These are regions that are called as "variant" based on cutoffs derived from a single sample that had high-depth (50x) WGS done and was then downsampled. These cutoffs are:

* Standard deviation of the segmented value matrix >= 0.08
* The number of probes used >= 67, covers 1Mb 
* Gains are called at >= 1.1
* Losses are called at <= 0.9


```{r var_reg, echo=F, message=F, warning=F}
prog.pt = subset(sum.patient.data, Status == 'P')$Hospital.Research.ID
np.pt = subset(sum.patient.data, Status == 'NP')$Hospital.Research.ID

var.seg.lengths = lapply(patient.data, function(df) {
  cbind('chr'=df$variable.regions$chrom, 'start'=df$variable.regions$start.pos, 'end'=df$variable.regions$end.pos, 'seg.len'=df$variable.regions$end.pos - df$variable.regions$start.pos)
})

segment.lengths = lapply(patient.data, function(df) {
  cbind('chr'=df$seg.vals$chrom, 'seg.len'=df$seg.vals$end.pos - df$seg.vals$start.pos)
})

sum.var = rbind(
  'Progressors'=do.call(rbind.data.frame, var.seg.lengths[prog.pt]) %>% summarise( 'n.segments' = length(seg.len), 'median.length'=median(seg.len), 'mean.length'=mean(seg.len), 'sd'=sd(seg.len)),
  'Non-Progressors'=do.call(rbind.data.frame, var.seg.lengths[np.pt]) %>% summarise( 'n.segments' = length(seg.len), 'median.length'=median(seg.len), 'mean.length'=mean(seg.len), 'sd'=sd(seg.len)))

pander(sum.var, caption='Variable regions between progressors and non-progressors')

sum.seg = rbind(
  'Progressors'=do.call(rbind.data.frame, segment.lengths[prog.pt]) %>% summarise( 'n.segments' = length(seg.len), 'median.length'=median(seg.len), 'mean.length'=mean(seg.len), 'sd'=sd(seg.len)),
  'Non-Progressors'=do.call(rbind.data.frame, segment.lengths[np.pt]) %>% summarise( 'n.segments' = length(seg.len), 'median.length'=median(seg.len), 'mean.length'=mean(seg.len), 'sd'=sd(seg.len)))


pander(sum.var, caption='Segment stats between progressors and non-progressors')
```


```{r echo=F, message=F, warning=F}

df = rbind(cbind(do.call(rbind.data.frame, var.seg.lengths[prog.pt]), 'status'='Progressor'), cbind(do.call(rbind.data.frame, var.seg.lengths[np.pt]), 'status'='Non-Progressor'))
lenWT = wilcox.test(subset(df, status == 'Progressor')$seg.len, subset(df, status == 'Non-Progressor')$seg.len)

ggplot(df, aes(status, seg.len)) + geom_jitter(aes(color=status), alpha=0.5, show.legend = F, width=.2) +
  geom_boxplot(aes(group=status, fill=status), show.legend=F, outlier.colour = NA) + 
  geom_label(data=subset(df, status == 'Progressor'), aes(status, max(seg.len)-2*sd(seg.len), label=paste('median\n',scientific(median(seg.len),3),sep='') )) +
  geom_label(data=subset(df, status == 'Non-Progressor'), aes(status, max(seg.len), label=paste('median\n',scientific(median(seg.len),3),sep=''))) +
  labs(y='Variable segment length', x='', title='Variable regions', caption=paste('Wilcox p-value', signif(lenWT$p.value, digits=3) ))

ggplot(df, aes(status, seg.len)) + 
  geom_jitter(aes(color=status), show.legend=F) + geom_violin(aes(group=status, fill=status), show.legend=F) +
  geom_label(data=subset(df, status == 'Progressor'), aes(status, max(seg.len)-2*sd(seg.len),label=paste('median\n',scientific( median(seg.len) , 3),sep='') ))  +
    geom_label(data=subset(df, status == 'Non-Progressor'), aes(status, max(seg.len), label=paste('median\n',scientific( median(seg.len) , 3),sep=''))) +
    labs(y='Variable segment length', x='', title='Variable regions', caption=paste('Wilcox p-value', signif(lenWT$p.value, digits=3) ))

```

```{r, echo=F, warning=F, message=F}
df = rbind(cbind.data.frame('n'=sapply(var.seg.lengths[prog.pt], nrow), 'status'='Progressor'), cbind.data.frame('n'=sapply(var.seg.lengths[np.pt], nrow), 'status'='Non-Progressor'))
nWT = wilcox.test(subset(df, status == 'Progressor')$n, subset(df, status == 'Non-Progressor')$n)

ggplot(df, aes(status, n, group=status)) + geom_jitter(aes(color=status), width=0.2) +
  geom_boxplot(aes(fill=status), position = 'dodge', show.legend = F, outlier.colour = NA) +
  geom_label(data=subset(df, status == 'Progressor'), 
             aes(status, max(n)-2*sd(n), label=paste('median\n',sprintf("%.1f",median(n)),sep=''), fill=status ), color='white' ) +
  geom_label(data=subset(df, status == 'Non-Progressor'), 
             aes(status, max(n)+2*sd(n), label=paste('median\n',sprintf("%.1f",median(n)),sep=''), fill=status), color='white') +
  labs(y='Number of segments', title="Total number of segments", caption=paste('Wilcox p-value', signif(nWT$p.value, digits=3)) ) +
  theme(legend.position = 'none')


ggplot(df, aes(status, n, group=status)) + geom_jitter(aes(color=status), width=0.2) +
  geom_violin(aes(fill=status, group=status)) +
  geom_label(data=subset(df, status == 'Progressor'), 
             aes(status, max(n)-2*sd(n), label=paste('median\n',sprintf("%.1f",median(n)),sep=''), fill=status ), color='white' ) +
  geom_label(data=subset(df, status == 'Non-Progressor'), 
             aes(status, max(n)+2*sd(n), label=paste('median\n',sprintf("%.1f",median(n)),sep=''), fill=status), color='white') +
  labs(y='Number of segments', title="Total number of segments", caption=paste('Wilcox p-value', signif(nWT$p.value, digits=3)) ) +
  theme(legend.position = 'none')

```


### Check variable segment length vs chromosome length

```{r echo=F, message=F, warning=F, fig.height=8, fig.width=20}
n.var.regions = sapply(var.seg.lengths, nrow)

#wilcox.test(n.var.regions[prog.pt], n.var.regions[np.pt])

sum.patient.data$Num.Var.Regions = as.vector(sapply(var.seg.lengths[sum.patient.data$Patient], nrow))

chr.info = read.table('hg19_chrom.txt',sep='\t', header=T)
chr.info$genome.length = cumsum(as.numeric(chr.info$length))

var.seg.lengths = lapply(var.seg.lengths, function(x) merge(x, chr.info, by='chr') )

# All segments
segment.lengths = lapply(segment.lengths, function(x) merge(x, chr.info, by='chr') )

# Genome coverage
genome.coverage = sapply(segment.lengths, function(x) {
  sum(as.numeric(x[['seg.len']]))/sum(as.numeric(chr.info$length))
})

# No difference in coverage
covtt = wilcox.test(genome.coverage[prog.pt], genome.coverage[np.pt])

var.seg.lengths = lapply(var.seg.lengths, function(x) cbind(x, 'chr.ratio'=x[['seg.len']]/x[['length']]))

# Segment lengths, variable regions
vsl = do.call(rbind.data.frame, lapply(patient.data, function(df) {
  if (is.null(df$variable.regions) || nrow(df$variable.regions) <= 0) return(NULL)
  
  mc = grep('^D', colnames(df$variable.regions))

  meanCN = sapply(apply(t(df$variable.regions[,mc]), 2, function(x) x[which(x >= 1.1 | x <= 0.9)]), mean)
  meanCN[is.na(meanCN)] = 0

    cbind.data.frame( 'chr'=df$variable.regions$chrom, 'start'=df$variable.regions$start.pos, 'end'=df$variable.regions$end.pos, meanCN, 'status'= df$info$Status[1])
  }))
vsl = vsl[-which(vsl$meanCN == 0),]

vsl$meanCN = vsl$meanCN - mean(vsl$meanCN)

vslm = melt(vsl, id.vars=c('chr','start','end','status'))

chr.info = subset(chr.info, chr %in% c(1:22))
chr.info$chr = ordered(chr.info$chr, levels=c(1:22))
vslm$chr =  ordered(vslm$chr, levels=c(1:22))
vslm = merge(chr.info, vslm, by='chr')
vslm = vslm %>% rowwise() %>% mutate('genome.start'=genome.length+start, 'genome.end'=genome.length+end)

vslm$status = ifelse(vslm$status == 'P', 'Progressor', 'Non Progressor')
  

ggplot(vslm, aes(x=length)) + 
  facet_grid(status~chr) +
  geom_rect(data=vslm, aes(xmin=start, xmax=end, ymin=0, ymax=value, fill=value)) +
  scale_fill_gradient(low = "green3", high = "red3", name='') + 
  labs(title="Variable region gains & losses", x='Chromosomes', y='Mean adjusted segmentation value') +
  theme(axis.text.x=element_blank(),axis.ticks.x=element_blank(), legend.position='bottom')



```

```{r echo=F, message=F, warning=F}
# Variable region length by ratio of chromosome
chr.ratios = lapply(var.seg.lengths, function(x) x[['seg.len']]/x[['length']])

melted = melt(chr.ratios)
melted = merge(sum.patient.data[,c('Hospital.Research.ID', 'Status')], melted, by.x='Hospital.Research.ID', by.y='L1')

chr.ratio.tt = t.test(unlist(chr.ratios[prog.pt]), unlist(chr.ratios[np.pt]))

ggplot(melted, aes(Status, value, fill=Status)) + geom_jitter(aes(color=Status), width=0.2) + geom_boxplot(outlier.colour = NA) + 
  geom_label(data=subset(melted, Status == 'P'), aes(x=Status, y=max(value)-sd(value), label=scientific(median(value),3) ), color='white', show.legend = F ) +
  geom_label(data=subset(melted, Status == 'NP'), aes(x=Status, y=max(value)-sd(value), label=scientific(median(value),3) ), color='white' , show.legend = F) +
  labs(title="Variable region length",subtitle=paste("p-value", signif(chr.ratio.tt$p.value, digits=3)), y='Segment Length/Chr length')

ggplot(melted, aes(Status, value, fill=Status)) + geom_jitter(aes(color=Status), width=0.2) + geom_violin() + 
  geom_label(data=subset(melted, Status == 'P'), aes(x=Status, y=max(value)-sd(value), label=scientific(median(value),3) ), color='white' , show.legend = F) +
  geom_label(data=subset(melted, Status == 'NP'), aes(x=Status, y=max(value)-sd(value), label=scientific(median(value),3) ), color='white' , show.legend = F) +
  labs(title="Variable region length",subtitle=paste("p-value", signif(chr.ratio.tt$p.value, digits=3)), y='Segment Length/Chr length')


var.seg.lengths =  lapply(var.seg.lengths, function(df) {
   ddply(df, .(chr, length), summarise, seg.lengths = sum(seg.len), chr.ratio = sum(seg.len)/length[1])
})


# Genome coverage of the variable regions
pcov = median(sapply(var.seg.lengths[prog.pt], function(x) 
  sum(as.numeric(x[['seg.lengths']]))/sum(as.numeric(x[['length']])) ), na.rm=T)

npcov = median(sapply(var.seg.lengths[np.pt], function(x) 
  sum(as.numeric(x[['seg.lengths']]))/sum(as.numeric(x[['length']])) ), na.rm=T)

## Grouping the chromosomes by length 1-12, 13-22
get.chrs<-function(x, chrs) { subset(x, chr %in% chrs)$chr.ratio }

long.chr.tt = t.test(unlist(lapply(var.seg.lengths[prog.pt], get.chrs, chrs=c(1:12))), unlist(lapply(var.seg.lengths[np.pt], get.chrs, chrs=c(1:12))))
short.chr.tt = t.test(unlist(lapply(var.seg.lengths[prog.pt], get.chrs, chrs=c(13:22))), unlist(lapply(var.seg.lengths[np.pt], get.chrs, chrs=c(13:22))))

#There may be an enrichment for segments in the shorter chromosomes amongst progressors but it's weak (fisher `r ftL$p.value`)
#df = data.frame( 'short' = rbind( 'np' = sum(table(np.regions$chrom)[1:12]),
#       'p' = sum(table(prog.regions$chrom)[1:12])),
#                 'long' = rbind( 'np' = sum(table(np.regions$chrom)[13:22], na.rm=T),
#       'p' = sum(table(prog.regions$chrom)[13:22], na.rm=T)))

#ftL = fisher.test(df)
```

So it appears that adjusting for chromosome length (looking at the ratio of the chromosome covered by segments rather than the raw bp) a significant difference in the coverage of the genome by CN segments that are variable between progressors and non: t-test p-value `r signif(chr.ratio.tt$p.value, 3)`.  Overall bp coverage of the sWGS is `r round(mean(genome.coverage), 2)*100`%  (P `r round(mean(genome.coverage[prog.pt]),2)`, NP `r round(mean(genome.coverage[np.pt]), 2)`).  

Variable regions cover `r round(pcov, 2)*100`% of the the progressors genome, and `r round(npcov, 2)*100`% of non-progressor genomes.

```{r cor, echo=F, message=F, warning=F, fig.height=10,fig.width=10, fig.align='left', eval=F}

lower<-function(rc) {
 ut = upper.tri( rc$r , diag=T) 
 rc$r[ut] = 0
 rc$P[ut] = NA
 return(rc)
}

sig.rc<-function(rc) {
  sig = rc$P > 0.05
  rc$r[sig] = 0
  return(rc)
}


cors = list()
for (name in names(patient.data)) {
  sample.normalised.segvals = patient.data[[name]]$norm.seg.vals # prog
  sample.normalised.segvals =  sample.normalised.segvals[, intersect(patient.data[[name]]$info$Samplename, colnames(sample.normalised.segvals)) ]
  
  info = subset(patient.data[[name]]$info, Samplename %in% colnames(sample.normalised.segvals))
  
  colnames(sample.normalised.segvals) =  paste(info$Endoscopy.Year,  c(1:nrow(info)) , sep='.')

  if (nrow(sample.normalised.segvals) <= 4) {
    warning(paste(name, 'cannot be run in rcorr, too few observations'))
  }
    
  #sample.normalised.segvals = patient.data$AD0361$norm.seg.vals # non-p
  rc = rcorr(as.matrix(sample.normalised.segvals), type='pearson')
  #rc = sig.rc(lower(rc))
  rc = sig.rc(rc)

  diff = unlist(apply(rc$r, 1, function(x)  sum(x[x < 0])-1 ))/nrow(rc$r)
  sim = unlist(apply(rc$r, 1, function(x)  sum(x[x > 0])-1 ))/nrow(rc$r)

  t.test(abs(sim), abs(diff))

  cors[[name]][['sim']] = sim
  cors[[name]][['diff']] = diff
}

unlist(lapply(cors, function(x) t.test(abs(x$sim), abs(x$diff))$p.value))

simPNP = wilcox.test(unlist(lapply(cors[subset(sum.patient.data, Status == 'P')$Patient], function(x) sum(x$sim))),
            unlist(lapply(cors[subset(sum.patient.data, Status == 'NP')$Patient], function(x) sum(x$sim))))

diffPNP = wilcox.test(unlist(lapply(cors[subset(sum.patient.data, Status == 'P')$Patient], function(x) sum(x$diff))),
            unlist(lapply(cors[subset(sum.patient.data, Status == 'NP')$Patient], function(x) sum(x$diff))))

#There's no clear difference (currently) between P and NP for the similar correlation (`r simPNP$p.value`) or the inverse (`r diffPNP$p.value`) correlations.  There are other ways we may be able to look at this still.

```

## Gain vs Loss

```{r echo=F, message=F, warning=F}

df = do.call(rbind.data.frame, lapply(patient.data, function(df) {
  nohgd = subset(df$info, !Pathology %in% c('HGD','IMC'))$Samplename
  gain=0;loss=0
  if ( !is.null(df[['variable.regions']]) ) {
    nohgd = intersect(nohgd, colnames(df$variable.regions))
    gain = length(which(apply(as.data.frame(df$variable.regions[,nohgd]), 1, function(x) length(which(x >= 1.1))) > 0))
    loss = length(which(apply(as.data.frame(df$variable.regions[,nohgd]), 1, function(x) length(which(x <= 0.9))) > 0))
  }
  
  cbind.data.frame('gain'=gain, 'loss'=loss, 'status'=unique(df$info$Status))
}))
head(df)

ttg = t.test(subset(df, status == 'P')$gain, subset(df, status == 'NP')$gain)
ttl = t.test(subset(df, status == 'P')$loss, subset(df, status == 'NP')$loss)

ft = fisher.test((df %>% group_by(status) %>% summarise( 'gains'=sum(gain), 'losses'=sum(loss)))[,c('gains','losses')])

ggplot(melt(df), aes(status, value, fill=status)) + facet_grid(.~variable) + geom_bar(stat='identity') +
  labs(title='Total gains & Losses excluding HGD/IMC', y='counts', x='', caption=paste('Gains and losses are signficantly enriched in Progressors vs Non.\nFisher odds ', round(ft$estimate,2), '(p=', signif(ft$p.value, digits=3), ')', sep='' )) 

```


## How do Cancer Gene Consensus genes map to the variable regions?

```{r, hmm, echo=F, message=F, warning=F, fig.height=60, fig.width=10}
ccgenes = read.table('~/Data/CosmicCensusGenes.tsv', sep='\t', header=T, stringsAsFactors=F)
ccgenes = ccgenes[-which(ccgenes$Mutation.Types == 'T'),]

noLoc = grep(':-',ccgenes$Genome.Location)
if (length(noLoc) > 0) {
  noLocGenes = ccgenes[noLoc,]
  ccgenes = ccgenes[-noLoc,]
}

locs = as.data.frame(do.call(rbind, strsplit(ccgenes$Genome.Location, ':|-')))
colnames(locs) = c('chr','start','end')
locs[c('start','end')] = lapply(locs[c('start','end')], function(x) as.numeric(as.character(x)))

ccgenes = cbind(locs, ccgenes)
ccgenes = makeGRangesFromDataFrame(ccgenes, keep.extra.columns = T, start.field = 'start', end.field = 'end')

# ov = findOverlaps(ccgenes, ccgenes)
# ov = ov[queryHits(ov) != subjectHits(ov)]
# ccgenes = ccgenes[-subjectHits(ov)[1]]

cancer.consensus.genes<-function(reg, ccgenes) {
  reg = as_tibble(reg)
  gr = GenomicRanges::makeGRangesFromDataFrame(reg, start.field = 'start.pos', end.field = 'end.pos')

  ov = findOverlaps(ccgenes, gr)
  reg$Gene.Symbols = NA
  for (hit in unique(subjectHits(ov))) {
    reg[hit, 'Gene.Symbols'] =  paste(ccgenes[queryHits(ov)[which(subjectHits(ov) == hit)],]$Gene.Symbol, collapse=',')
  }
  
  rwg = cbind(reg, reg[,-(1:5)])
  return(rwg)
}

rm(var.regions)
for (patient.name in names(patient.data)) {
  regions = patient.data[[patient.name]]$variable.regions[,c('chrom','arm','start.pos','end.pos')]
  if (is.null(regions) || nrow(regions) <= 0) next

  reg = cancer.consensus.genes(regions, ccgenes)
  x = patient.data[[patient.name]]$variable.regions[which(!is.na(reg$Gene.Symbols)), intersect(patient.data[[patient.name]]$info$Samplename, colnames(patient.data[[patient.name]]$variable.regions))  ]

  gl = apply(x, 1, function(y)  ifelse(length(which(y >= 1.1)) > length(which(y <= 0.9)), 'gain', 'loss') )
  reg[which(!is.na(reg$Gene.Symbols)), 'CN'] = gl
  reg = reg[,c('chrom','arm','start.pos','end.pos','CN','Gene.Symbols')]

  for (i in 1:nrow(reg)) {
    if (is.na(reg[i,'Gene.Symbols'])) next
    genes = unlist(strsplit(reg[i,'Gene.Symbols'], ','))
    
    reg[i,'Gene.Symbols'] = genes[1]
    
    if (length(genes) > 1)
      reg = rbind(reg, cbind(reg[i, (1:5)], 'Gene.Symbols' = genes[2:length(genes)]))
  }
  reg = arrange(reg, chrom, arm, start.pos)
  
  status = subset(sum.patient.data, Hospital.Research.ID == patient.name)$Status
  
  reg = cbind(patient.name, status, reg)
    
  if (!exists('var.regions')) {
    var.regions = reg
  } else {
    var.regions = rbind(var.regions, reg)
  }
}

var.regions = var.regions[!is.na(var.regions$Gene.Symbols),]
var.regions$patient.name = as.character(var.regions$patient.name)

var.regions = var.regions[-grep('CDKN2A\\(p14\\)', var.regions$Gene.Symbols),]

gene.cnt = ddply(var.regions, .(Gene.Symbols, CN), summarise, Progressors=length(which(status == 'P')), Non.Progressors=length(which(status == 'NP')))
gene.cnt[,unique(var.regions$patient.name)] = 0
gene.cnt = gene.cnt[which(!is.na(gene.cnt$Gene.Symbols)),]

for (i in 1:nrow(gene.cnt)) {
  gene = as.character(gene.cnt[i,'Gene.Symbols'])
  type = as.character(gene.cnt[i,'CN'])
#  print(paste(gene, type))
#  print(subset(var.regions, Gene.Symbols == gene & CN == type)$patient.name)
  gene.cnt[i,unique(subset(var.regions, Gene.Symbols == gene & CN == type)$patient.name)] = 1
}

gene.cnt = arrange(gene.cnt, -Progressors)
## RAD51B is screwing up the plot simply because it shows 16 non-progressors (8 progressors) with a gain but the next closest is 9 so we'll just filter that out for now
#gene.cnt = subset(gene.cnt, gene.cnt$Gene.Symbols != 'RAD51B')

#gene.p = subset(gene.cnt, Progressors >= Non.Progressors)
gene.p = gene.cnt

gene.p = gene.p[,(1:4)]
m = melt(gene.p, id.vars=c('Gene.Symbols', 'CN'))
m = transform(m, Gene.Symbols = reorder(Gene.Symbols, value))
m$CN = factor(m$CN)

m$brks <- cut(m$value, include.lowest=T, right=F,ordered_result=T,
                   breaks=c(0, 1, 3, 5, 8, 10, 12, 15),
                   labels=c('0','1-3', '4-5', '6-8','9-10', '10-12', '>12'))

## TODO: Tile the variables so that you can see the gains/losses per progressor/non
#subset(m, Gene.Symbols %in% c('CDKN2A', 'TBL1XR1'))
m = transform(m, Gene.Symbols = reorder(Gene.Symbols, brks))

myP = c('white',brewer.pal(n = length(levels(m$brks)), name = "Reds"))

ggplot(m,  aes(y=Gene.Symbols, x=CN, fill=brks)) + geom_tile() + facet_grid(.~variable) + 
    scale_fill_manual(values = myP, 'Patient Counts') + labs(title="All available COSMIC genes", x='', y='')

m2 = melt(gene.p %>% group_by(Gene.Symbols) %>% summarise( 'Progressors' = sum(Progressors), 'Non.Progressors' = sum(Non.Progressors)), id.vars='Gene.Symbols')

m2 = transform(m2, Gene.Symbols = reorder(Gene.Symbols, value))

m2$brks <- cut(m2$value, include.lowest=T, right=F,ordered_result=T,
                   breaks=c(0, 1, 3, 5, 8, 10, 12, 15),
                   labels=c('0','1-3', '4-5', '6-8','9-10', '10-12', '>12'))

ggplot(m2,  aes(y=Gene.Symbols, x=variable, fill=brks)) + geom_tile() + #facet_grid(.~variable) + 
    scale_fill_manual(values = myP, 'Patient Counts') + labs(title="All available COSMIC genes", x='', y='')  
  
head(m2)
m2

print.png<-function(p, name, ...) {
  png(filename=name,...)
  print(p)
  dev.off()
}


pg_genes = subset(m2, variable == 'Progressors')
pg_genes = pg_genes[order(-pg_genes$value),]

limits = seq(1, 220, 50)
for (i in limits) {
  max = i+49
  if (max > nrow(pg_genes) ) max = nrow(pg_genes)
  
  p = ggplot(subset(m2, Gene.Symbols %in% pg_genes$Gene.Symbols[c(i:max)]),  aes(y=Gene.Symbols, x=variable, fill=brks)) + geom_tile() +
    scale_fill_manual(values = myP, 'Patient Counts') +
    labs(title="All available COSMIC genes", x='', y='', subtitle=paste("genes ", i, "-", max, " by number of patients", sep='')) +
    theme(text=element_text(size=16, face="bold"), axis.title=element_text(size=18))
  
  if (i != 1)
    p = p + theme(legend.position = 'none')
    
  print.png(p, paste('cosmic_genes_', i, '-', max, '.png', sep=''), width=600, height=1000)
}

```

Just the top 50 genes by the number of patients that share variants in these regions.

```{r echo=F, warning=F, message=F, fig.height=12, fig.width=8}
gc = as.data.frame(gene.p %>% group_by(Gene.Symbols) %>% summarise(total = sum(Progressors) + sum(Non.Progressors)))
gc = transform(gc, Gene.Symbols = reorder(Gene.Symbols, total))
gc = arrange(gc, -total)

m = melt(subset(gene.p, Gene.Symbols %in% gc[1:50, 'Gene.Symbols']), id.vars=c('Gene.Symbols', 'CN'))
m = transform(m, Gene.Symbols = reorder(Gene.Symbols, value))

m$brks <- cut(m$value, include.lowest=T, right=F,
                   breaks=c(0, 1, 3, 5, 8, 10, 12, 15),
                   labels=c('0','1-3', '4-5', '6-8','9-10', '10-12', '>12'))

ggplot(m,  aes(y=Gene.Symbols, x=CN, fill=brks)) + geom_tile(color='white') + facet_grid(.~variable) + 
    scale_fill_manual(values=myP, 'Patient Counts') + 
    labs(title="Top 50 genes by number of patients", x='', y='')



gene.p50 = gene.p %>% group_by(Gene.Symbols) %>% summarise( 'Progressors' = sum(Progressors), 'Non.Progressors' = sum(Non.Progressors))

m2 = melt(subset(gene.p50, Gene.Symbols %in% gc[1:50, 'Gene.Symbols']), id.vars=c('Gene.Symbols'))

m2 = transform(m2, Gene.Symbols = reorder(Gene.Symbols, value))

m2$brks <- cut(m2$value, include.lowest=T, right=F, ordered_result=T,
                   breaks=c(0, 1, 3, 5, 8, 10, 12, 15),
                   labels=c('0','1-3', '4-5', '6-8','9-10', '10-12', '>12'))

ggplot(m2,  aes(y=Gene.Symbols, x=variable, fill=brks)) + geom_tile(color='white') + 
    scale_fill_manual(values=myP, 'Patient Counts') + 
    labs(title="Top 50 genes by number of patients", x='', y='')

#ggplot(m, aes(y=Gene.Symbols, x=variable)) + 
#  geom_tile(aes(fill=brks), color='white') + scale_fill_manual( values = myP, "Patient Counts") +
#  geom_point(data=subset(m, value > 0), aes(y=Gene.Symbols, x=variable, shape=CN, color=CN), size=3 ) +
#  scale_color_manual("CN", values=c("blue", "yellow")) + scale_shape_manual(values=c(17,20)) +
#  labs(title="Top 50 genes by number of patients", x='', y='')
```


# Is HGD over represented for variable regions?

It's possible that the variable regions we find are entirely within the HGD/IMC samples.  So are they overrepresented?

```{r, hgd, echo=F, message=F, warning=F}

pvnp = data.frame(matrix(nrow=length(subset(sum.patient.data, Status == 'P')$Hospital.Research.ID), ncol=2, dimnames=list(subset(sum.patient.data, Status == 'P')$Hospital.Research.ID, c('hgd','be'))))    
for (name in subset(sum.patient.data, Status == 'P')$Hospital.Research.ID) {
  var.regions = patient.data[[name]]$variable.regions
  if ( is.null(var.regions) || nrow(var.regions) == 0 ) next
  
  sn = which(colnames(patient.data[[name]]$variable.regions)[-(1:5)] %in% subset(patient.data[[name]]$info, grepl('HGD|IMC', Pathology))$Samplename)
  if (length(sn) <= 0 | nrow(patient.data[[name]]$variable.regions) <= 0) next
  
  ## HGD
  pvnp[name, 'hgd'] = length(which(patient.data[[name]]$variable.regions[,sn] >= gain.threshold | 
                                patient.data[[name]]$variable.regions[,sn] <= loss.threshold))
  
  # BE-LGD
  pvnp[name, 'be'] = length(which(patient.data[[name]]$variable.regions[-sn] >= gain.threshold | 
                                patient.data[[name]]$variable.regions[-sn] <= loss.threshold))
}

## TODO Check enrichment of variable regions found in BE vs HGD 

#tt = t.test(pvnp$hgd, pvnp$be)
pvnp = cbind(pvnp, (pvnp %>% rowwise() %>% mutate(total=hgd+be, hgd.ratio=round(hgd/(hgd+be),3) ))[c('total','hgd.ratio')])

ft = fisher.test( pvnp[complete.cases(pvnp[,c('hgd','be')]), c('hgd','be')], simulate.p.value = T)

ct = cor.test(subset(sum.patient.data, Status == 'P')$total.samples, pvnp$hgd.ratio)

pander(pvnp, justify='left', 
       caption=paste("Number of variable regions found in HGD vs non-HGD samples of progessor patients.  There's a significant difference between them (Fisher's p=",signif(ft$p.value,3),').  Not all patients have HGD samples (NAs). There is no significant correlation (r= ',signif(ct$estimate, 2),') between the number/ratio of variable regions found in HGD samples vs non (Pearsons p=',signif(ct$p.value, 3),')',  sep=''))

```


```{r, warning=F, eval=F}
#This is just looking at the distribution of values per sample. The average ploidy per sample is around 1.  Currently we're calling variable regions based on this.
ks = list()
for (pt in names(patient.data)) {
  print(pt)
  cols = intersect(arrange(patient.data[[pt]]$info, Pathology, Endoscopy.Year)$Samplename, grep('^D\\d+', colnames(patient.data[[pt]]$seg.vals), value=T))
  if (length(cols) <= 1) next
  
  segvals = patient.data[[pt]]$seg.vals[,cols]
  
  sdx = round(apply(segvals, 2, sd), 3)
  se = round(apply(segvals, 2, function(x) sd(x)/sqrt(length(x))), 3)

 gg = ggplot(melt(segvals)) + 
        facet_wrap(~variable,scales = "free_x") + 
        geom_histogram(bins=10, fill='blue', color='grey',aes(value, ..density..)) +
        geom_density(aes(value+sdx/2), color='red', size=1) +
        geom_density(aes(value-sdx/2), color='green', size=1) +
        labs(title=paste(pt, '  (',patient.data[[pt]]$info$Status[1],')', sep=''))
  print(gg)   
  #plots[[pt]] = gg
  
  ks[[pt]] = sapply(segvals, function(x) ks.test(x+sdx, x-sdx)$statistic  )
  print(pt)
  print(colSums(abs(round(patient.data[[pt]]$seg.vals[,-(1:5)], 1)-1)))
}




```


